core/src/data_sources/data_source.rs

```diff
@@ -392,9 +393,51 @@ impl DataSource {
         tags: &Vec<String>,
         source_url: &Option<String>,
         text: &str,
+        preserve_system_tags: bool,
     ) -> Result<Document> {
+        // disallow preserve_system_tags=true if tags contains a string starting with the system tag prefix
```

Comment:
Maybe add a sentence to the comment: this guarantees that we don't have duplicate system tags when preserving system tags?

Reply:
Adding ! Also, it prevents people from upserting with `__DUST*` tags from the API / UI